### ISSUE TITLE
PWA 2021

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/almanac.httparchive.org#2153]
# Part II Chapter 16: PWA

![PWA illustration](https://raw.githubusercontent.com/HTTPArchive/almanac.httparchive.org/main/src/static/images/2019/pwa/hero_lg.jpg)

If you're interested in contributing to the PWA chapter of the 2021 Web Almanac, please reply to this issue and indicate which role or roles best fit your interest and availability: [author](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide), [reviewer](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Reviewers'-Guide), [analyst](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide), and/or [editor](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Editors'-Guide).

## Content team

| Lead | Authors | Reviewers | Analysts | Editors | Coordinator |
| --- | ------- | --------- | -------- | ------- | -- |
| @demianrenzulli | @demianrenzulli | @webmaxru @Schweinepriester @thepassle @hemanth @tropicadri @andreban @jeffposnick  | @tunetheweb @demianrenzulli | - | @obto |

<details>
<summary><b>Expand for more information about each role</b></summary>

- The **[content team lead](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Content-Team-Leads'-Guide)** is the chapter owner and responsible for setting the scope of the chapter and managing contributors' day-to-day progress.
- **[Authors](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide)** are subject matter experts and lead the content direction for each chapter. Chapters typically have one or two authors. Authors are responsible for planning the outline of the chapter, analyzing stats and trends, and writing the annual report.
- **[Reviewers](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Reviewers'-Guide)** are also subject matter experts and assist authors with technical reviews during the planning, analyzing, and writing phases.
- **[Analysts](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide)** are responsible for researching the stats and trends used throughout the Almanac. Analysts work closely with authors and reviewers during the planning phase to give direction on the types of stats that are possible from the dataset, and during the analyzing/writing phases to ensure that the stats are used correctly.
- **[Editors](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Editors'-Guide)** are technical writers who have a penchant for both technical and non-technical content correctness. Editors have a mastery of the English language and work closely with authors to help wordsmith content and ensure that everything fits together as a cohesive unit.
- The **[section coordinator](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Section-Leads'-Guide)** is the overall owner for all chapters within a section like "User Experience" or "Page Content" and helps to keep each chapter on schedule.

_Note: The time commitment for each role varies by the chapter's scope and complexity as well as the number of contributors._

For an overview of how the roles work together at each phase of the project, see the [Chapter Lifecycle](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle) doc.
</details>

### [**0. Form the content team**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#0-create-content-team)
+ [x] **May 31**: The content team has at least one author, reviewer, and analyst

### [**1. Plan content**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#1-plan-content)
+ [x] **June 15** The content team has completed the chapter outline in the [draft doc][~google-doc]

### [**2. Gather data**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#2-gather-data)
+ [x] **June 30**: Analysts have added all necessary [custom metrics](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/README.md) and drafted a PR ([example](https://github.com/HTTPArchive/almanac.httparchive.org/pull/1087)) to track query progress
- **July 1 - 31**: HTTP Archive runs the July crawl

### [**3. Validate results**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#3-validate-results)
+ [x] **September 30**: Analysts have queried all metrics and saved the output to the [results sheet][~google-sheets]

### [**4. Draft content**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#4-draft-content)
+ [ ] **October 31**: The content team has written, reviewed, and edited the chapter in the [doc][~google-doc]

### [**5. Publication**](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#5-publication)
+ [ ] **November 15**: The completed chapter and all required [metadata](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide#metadata-to-add-at-the-top-of-your-chapters) and [figures](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Figures-Guide) are converted to markdown and submitted to [GitHub][~chapter-markdown]
- **December 1**: Target launch date 🚀

## Chapter resources

Refer to these 2021 PWA resources throughout the content creation process:

[📄 Google Docs][~google-doc] for outlining and drafting content
[🔍 SQL files][~sql-files] for committing the queries used during analysis
[📊 Google Sheets][~google-sheets] for saving the results of queries
[📝 Markdown file][~chapter-markdown] for publishing content and managing public metadata

[~google-doc]: https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit?usp=sharing
[~sql-files]: https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2021/pwa
[~google-sheets]: https://docs.google.com/spreadsheets/d/16AkIdDBBkCR5Kgb7kyfYvnNLQBu23Vsh7MUSFHW9RtA/edit?usp=sharing
[~chapter-markdown]: https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/src/content/en/2021/pwa.md

**Comment by webmaxru** [comment]
It's my pleasure to support PWA chapter as a reviewer!

**Comment by rviscomi** [comment]
Welcome @webmaxru!

**Comment by Schweinepriester** [comment]
If I may, I'd support PWA as a reviewer as well :)

**Comment by demianrenzulli** [comment]
I can work as an author in this one! 

cc // @andreban @petele @b1tr0t

**Comment by rviscomi** [comment]
@demianrenzulli thanks for your interest in authoring this chapter! As the content team lead, you'll be responsible for the scope and direction of the chapter and keeping it on schedule. We automatically monitor the staffing and progress of each chapter based on the state of the [initial comment][~chapter-metadata] so please keep that updated as you add new contributors and meet each milestone.

We've created a [Google Doc][~google-doc] for this chapter, which you're encouraged to use to collaborate with the content team on the initial outline, metrics, and ultimately the final draft.

Next steps for this chapter are:

- Due **May 31**: Complete [Milestone 0: Form the content team](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#0-create-content-team) by finding someone to be an analyst
- Due **June 15**: Get started on [Milestone 1: Plan content](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Chapter-Lifecycle#1-plan-content) by brainstorming some content ideas in the [doc][~google-doc] (you may need to request edit access, cc @webmaxru @Schweinepriester)

@obto will be the section coordinator for this chapter, so they'll be periodically checking in with you directly to make sure the chapter is staying on schedule. Reach out to them here in this issue if you have any questions about the process.

More information about the [content team lead](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Content-Team-Leads'-Guide) and [author](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide) roles and responsibilities are available for reference in the wiki if needed.

To anyone else interested in contributing to this chapter, please comment below to join the team!


[~chapter-metadata]: https://github.com/HTTPArchive/almanac.httparchive.org/issues/2153#issue-868436069
[~google-doc]: https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit?usp=sharing

**Comment by rviscomi** [comment]
📟 paging 2019/2020 contributors: @tomayac @jeffposnick @logicalphase @ahmadawais @jrharalson @hemanth @thepassle @jadjoubran @pearlbea @gokulkrishh @jaisanth @bazzadp

Would any of you be interested to contribute to the 2021 chapter? This chapter could use your help with reviewing and/or analyzing. It'd be great to have your support!

**Comment by thepassle** [comment]
Yeah I'd be happy to help out as reviewer where possible :)

**Comment by hemanth** [comment]
Count me in for reviews, thank you!

**Comment by demianrenzulli** [comment]
Thanks for your interest folks!

It seems like @bazzadp acted as analyst for the 2020 section. 

Hi Barry! Nice to meet you. You're more than welcome to collaborate on that role once again, but if you can't make it this year, and we don't find another contributor willing to work as analyst, I would be happy to take that role as well. 
In that case, maybe we can touch base separately, so you can share some insights on your experience with the 2020 PWA section?

We might be able to reuse several of the [2020 queries](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2020/pwa), plus other ideas we have in mind.

**Comment by demianrenzulli** [comment]
For the rest of the reviewers: it's the first time I contribute, so I'm not sure how you interacted in the past, but I'll touch base with Rick and other contributors to see how we can get organized for this year's PWA section 👍👍👍

**Comment by bazzadp** [comment]
Hey @demianrenzulli , I'm one of the leads her for the Web Almanac project. I stepped in as the analyst last year as we were short, but wouldn't say PWAs are my area of expertise so would be great if someone else could take the mantle here. But definitely more than happy to help guide them and help out where I can.

A lot of the queries was based on @tomayac 's work from 2019, and in particular [this blog post of his](https://medium.com/dev-channel/progressive-web-apps-in-the-http-archive-614d4bcf81fe) that he used as the based of that first year's chapter. See also https://github.com/HTTPArchive/almanac.httparchive.org/issues/1258#issuecomment-689857902 for more info on regenerating that data for last year (we'll help get that data for you again assuming you want it). Not sure if you wanna chip in here @tomayac ? Though know you are involved in Capabilities chapter too.

> I'll touch base with Rick and other contributors to see how we can get organized for this year's PWA section 

Best thing is for all authors to open [the draft word doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit?usp=sharing), request edit permission, and then start listing ideas of what metrics or sections you'd like to keep from 2019 and 2020, and what you'd like to introduce. Start to get a chapter outline based on that and comment and suggest things. more ideas the better at the beginning, and then can see which ones are realistic and whittle it down. The earlier you start this the less pressure you'll feel later :-)

And btw you can get links to the results sheets (and the queries) from 2019 and 2020 at the bottom of the chapters to see what data those authors had to play with, to help with that:
- 2019: https://almanac.httparchive.org/en/2019/pwa#explore-results
- 2020: https://almanac.httparchive.org/en/2020/pwa#explore-results

**Comment by tomayac** [comment]
> A lot of the queries was based on @tomayac 's work from 2019, and in particular [this blog post of his](https://medium.com/dev-channel/progressive-web-apps-in-the-http-archive-614d4bcf81fe) that he used as the based of that first year's chapter. See also [#1258 (comment)](https://github.com/HTTPArchive/almanac.httparchive.org/issues/1258#issuecomment-689857902) for more info on regenerating that data for last year (we'll help get that data for you again assuming you want it). Not sure if you wanna chip in here @tomayac ? Though know you are involved in Capabilities chapter too.

I've signed up for capabilities and am reluctant to overcommit. More than happy to answer ad-hoc questions you may have, though.

**Comment by demianrenzulli** [comment]
Thanks a lot @bazzadp!! If, based on previous experiences you think it's "doable" for someone to be both the author and analyst, I would be happy to collaborate on both sides.

As mentioned, this is my first time contributing, so yours, @tomayac and other folk's guidance, would be highly appreciated.

There are a bunch of resources here that I need to process before moving to the next step.

I'll keep this issue updated.
If anyone has any thoughts to share, please, let me know!

**Comment by rviscomi** [comment]
It's definitely doable to be an author and analyst, you wouldn't be the first. Each role peaks in activity at different times during the project, so it should be manageable for a single person assuming you can set aside a few hours each week.

**Comment by obto** [comment]
Hey @demianrenzulli excited to work with you and the rest of the group this year on the Almanac. I'm your go to guy if you've got any questions or need help so don't hesitate to reach out to me on github, the [Slack](https://join.slack.com/t/httparchive/shared_invite/zt-45sgwmnb-eDEatOhqssqNAKxxOSLAaA) (@obto) or email (david@davidjfox.com).

Few first steps:

1. All authors, reviewers and analysts should add themselves to the [Google Doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit) in the format: `Name (email@example.com)`. But looks like you've already got a good handle on that.
2. I've added links within the doc to the previous years Google doc in case you'd like to mine it for ideas.
3. Would love to set up a 30 minute Zoom call in the next couple weeks to kick-start the chapter planning and brainstorming process, and put some faces to the names of the people we'll all be working with this year. I'll reach out again later this week to find a time that works.

Excited to work with you all this year.

**Comment by demianrenzulli** [comment]
This is great @obto! Nice to meet you too.

My apologies for the delayed response, but we had a couple of things for I/O week, and some articles I've been working on for Web.dev.

I'm planning to start working on this by tomorrow. I'll make sure to also touch base with the rest of the collaborators.

We had some internal conversations already, and would be great to hear thoughts from the rest of the folks in this group.

I hope we can create a great PWA chapter this year!

**Comment by thepassle** [comment]
> All authors, reviewers and analysts should add themselves to the Google Doc in the format: Name (email@example.com). But looks like you've already got a good handle on that.

Looks like I need some rights to be able to add myself there, I've requested them in the doc.

**Comment by obto** [comment]
How does Tuesday (May 25) 11a ET / 8a PT / 4p BST work for that 30m chat? Timezones here: https://www.timeanddate.com/worldclock/meetingdetails.html?year=2021&month=5&day=25&hour=15&min=0&sec=0&p1=64&p2=179&p3=224&p4=136&p5=37&iv=1800

@demianrenzulli @webmaxru @Schweinepriester @thepassle @hemanth

**Comment by logicalphase** [comment]
Good for me.

**Comment by demianrenzulli** [comment]
That time: next Tuesday 11ET, works for me! Thanks for organizing this @obto!

**Comment by webmaxru** [comment]
Works fine for me!

**Comment by obto** [comment]
@demianrenzulli @logicalphase can you update the [Google doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit) with your email? I'll send the calendar invite to that email

**Comment by demianrenzulli** [comment]
@obto I just requested edit access with my Google account. At this point, since we don't have any other volunteer for analyst, I'll be signing up for that role as well, but if anyone is interested on taking on that role, that would be welcome.

**Comment by demianrenzulli** [comment]
Also, @andreban: would you like to sign up as reviewer by adding your details to the [Google doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit). We have a kick-off call on Tuesday, so, let us know if you want to participate.

**Comment by obto** [comment]
@demianrenzulli Added you to the analyst role, but we'll keep the `help wanted: analysts` tag and see if we cant find you one

**Comment by demianrenzulli** [comment]
That's great! I'd be happy to collaborate with analyst if you know about any that might be interested. This is my first time contributing to Web Almanac, so playing both the author and analyst role might be a bit too ambitious. 
In any case, from Ricks' comments it is totally doable, so I might be able to secure help from previous analysts to move it forward on my own 👍

**Comment by Schweinepriester** [comment]
That time works for me as well! Thanks for the link, as I'm in CEST! :)

**Comment by logicalphase** [comment]
I thought I had access to the Google doc, but just put a request for
access. I'll update soon as I can.

**Comment by obto** [comment]
Can do. Just updated the calendar invite to include everyone who's requested access to the google doc so far

**Comment by demianrenzulli** [comment]
I've just added @tropicadri to the doc and I see she has been already added to Tuesday's sync.

I'm not very sure of what would be the main topic of discussion for Tuesday, but, just in case, I found it useful to read the [2019](https://almanac.httparchive.org/en/2019/pwa) and [2020](https://almanac.httparchive.org/en/2020/pwa) editions of the Web Almanac to have an idea of what things were covered in the past. If you have the time, it might be a good idea to do it before the call to have all the context.

Here are also the [2019](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2019/pwa) and [2020](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2020/pwa) queries if you want to take a look.

I added a [Team's Notes](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit#heading=h.4fleu355f29s) section in the doc, where you can find some ideas I and others have been discussing recently.

Please, feel free to add any ideas or comments you might have as well! It would really help to count with different perspectives around these topics.

See you on Tuesday!

**Comment by obto** [comment]
@demianrenzulli Just sent out an agenda for the call on Tuesday. Looking forward to meeting you all 😄

**Comment by obto** [comment]
@demianrenzulli Can you tick the first milestone checkbox above `0. Form the content team`? Keeping these milestones up to date helps to give me an overview on how the Almanac as a whole is coming together

**Comment by demianrenzulli** [comment]
Just did. Thanks for the reminder @obto!

**Comment by obto** [comment]
@demianrenzulli Here's last years github issue, where you can find all the data they used https://github.com/HTTPArchive/almanac.httparchive.org/issues/909

**Comment by tomayac** [comment]
> Parse the Service Worker URL for any events to replace the events query query. Is this sufficient or can a service worker URL load other dependencies? The way we've done it in the past has always struck me as a bit brittle because of that.

Service workers can `importScripts()`, so strictly speaking you'd have to consider those.

**Comment by demianrenzulli** [comment]
Adding @jeffposnick, who might be able to validate my answers around service workers. Other folks in this issue, feel free to cime in as well:

> 1. Identify whether this is a SW worker page (is checking for the presence of a navigator.serviceWorker.register line either in HTML or any of the JS resources the best way of doing this or is there something cleverer?) - true/false.

I'm not aware of any other API to register service workers., so It seems like checking for that one would cover most scenarios. I see some cases where this might not work though (for example, when using [Workbox Window](https://developers.google.com/web/tools/workbox/modules/workbox-window)) and calling: `const wb = new Workbox('/sw.js');   wb.register(); `.
I assume that the library calls `navigator.serviceWorker.register` internally, so, if we can inspect calls made inside libraries, that would work as well?

> 2. Can you register SWs for other domains (e.g. iframes or other third-party loads)? If so would be good to track 1st party/3rd party too.

[This article](https://adactio.com/journal/10924) explains how to use that technique to register service workers for 3rd-party origins using iframes. I believe that's also what the [amp-install-serviceworker](https://amp.dev/documentation/examples/components/amp-install-serviceworker/) component used to do internally, in order to allow AMP Pages served from the Google AMP Cache to register the service worker that belongs to the publisher's origin. 

In any case, I don't think this works across browsers (I'm pretty sure that it doesn't work in Safari). With that said, I don't know how frequent this technique might be.

> 3. The Service Worker URL from the navigator.serviceWorker.register line, where true (could there be multiple service workers per page?).

There can be, at most, one service worker controlling a given page at the time, but I believe that there are no restrictions for a page to have multiple calls to `navigator.serviceWorker.register`, as long as it indicates for which scope that service worker has to be registered.

> 4. Identify whether this is has a manifest (presence of a <link rel="manifest"...> in the main HTML document I presume?) - true/false.

Detecting the presence of the Manifest sounds like a good way to check if sites could potentially be "installable". I'm seeing in 2020 there was also a [Lighthouse query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/pwa/lighthouse_pwa_audits.sql) to detect the % of installable manifests.

Besides that, if you are thinking on using this as a way to detect if a site is a PWA, I think it really depends on which definition of what is a PWA we end up adopting for this year's Almanac. In 2020, the definition seem to have been strongly based on the presence of the service worker:

__"The crux of a progressive web app is the service worker, which can be thought of as a proxy sitting between the browser and user. A service worker gives the developer total control over the network, rather than the network controlling the application."__

> 5. Parse the Service Worker URL for any events to replace the events query. Is this sufficient or can a service worker URL load other dependencies? The way we've done it in the past has always struck me as a bit brittle because of that. Any thoughts @tomayac ?

I think Thomas already replied this one. I don't have more to add to that question, but I'm leaving this one open for @jeffposnick to chime in.

@bazzadp I hope I have understood the questions correctly and the motivations behind them. I would be happy to have a call with you and other folks in this issue if you think that's a good idea.

**Comment by bazzadp** [comment]
> Besides that, if you are thinking on using this as a way to detect if a site is a PWA, I think it really depends on which definition of what is a PWA we end up adopting for this year's Almanac. In 2020, the definition seem to have been strongly based on the presence of the service worker:

Yes, I think that's key! We did switch definitions last year:

- 2019 - Presence of Manifest **or** Service Worker
- 2020 - Presence of Manifest **and** Service Worker

What's your thoughts for this year?

**Comment by demianrenzulli** [comment]
Thanks for confirming that @bazzadp! It's great to see that this was actually a topic that came out in the past.
Coming up with a new definition of what a PWA is in 2021, is one of the main AIs that came up from our initial sync.

The problem I see with the 2020 definition:  "Presence of Manifest and Service Worker", is that we have seen websites that have both things and don't do much with them. For example, they don't end up prompting the user to install the site and/or they can have an empty or very simple service worker that don't do much for the UX.
On the other hand, there are sites that might not have either a manifest or a service worker, but they still provide and "App-like" UX that would probably be a better candidate to be called a "PWA" than those that don't.

One idea that I had to solve this problem would be to list properties that sites usually have to be considered a PWA (similar to what [was done in the original definition](https://infrequently.org/2015/06/progressive-apps-escaping-tabs-without-losing-our-soul/)), and then say that we'll be presenting stats of adoption of different features (service workers, manifests, etc.). If a site combines many of them, then the end experience might be much powerful, but some of them only use some of these features. 

This granular way of presenting features seems to be aligned with what was done in 2020 (please, correct me if I'm wrong).

Now, if we really need to come up with a way of detecting PWAs out there for query purposes, I agree the 2020 definition is the best we have.

As said, we have just started discussing this, so any thoughts are highly welcome.
Cheers.

**Comment by bazzadp** [comment]
We also have access to all the [Lighthouse PWA audits](https://web.dev/lighthouse-pwa/) so could define the list based on some of those? Or the PWA score calculated by Lighthouse?

Only word of caution is that if we change the definition (again!) this year then we should make it clear in the chapter so people don't try to compare previous chapter. We called that out [in last year's chapter](https://almanac.httparchive.org/en/2020/pwa#:~:text=This%20is%20different%20than%20the%20approach%20taken%20in%20last%20year%E2%80%99s%20PWA%20chapter%20which%20looked%20at%20overall%20manifest%20usage%2C%20so%20you%20may%20notice%20some%20differences%20in%20results%20this%20year.):

> Having a web app manifest does not necessarily indicate the site is a progressive web app, as they can exist independently of service worker usage. However, as we are interested in PWAs in this chapter, we have investigated only those manifests for sites where a service worker also exists. This is different than the approach taken in last year’s PWA chapter which looked at overall manifest usage, so you may notice some differences in results this year.

**Comment by demianrenzulli** [comment]
Taking Lighthouse into account would be a good idea, since it's the only tool I'm aware of that provides a badge to sites that considers a PWA according to a set of attributes. Now, since Lighthouse changes frequently, let me check internally how up-to-date [this list](https://web.dev/lighthouse-pwa/) might be. I believe that, if we update [this section](https://almanac.httparchive.org/en/2020/pwa#lighthouse-insights) with the new checks that Lighthouse makes we should be fine, and that would be aligned with the new definition we want to provide in the [Introduction section](https://almanac.httparchive.org/en/2020/pwa#introduction).

Regarding the stats about Manifests, I like the approach you've mentioned that was taken [for last year edition](https://almanac.httparchive.org/en/2020/pwa#web-app-manifests). I wouldn't change that part.

@bazzadp I hope I have provided some clarity, but we are still in the early stages. As said, if you are open to jump into a call next week, we could definitely go over all these points.

Just let me know.

**Comment by demianrenzulli** [comment]
@jeffposnick just confirmed that the responses provided [before](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2153#issuecomment-849760433) are all correct. He added this:

> The one addition would be regarding <iframe>-based service worker registrations. Those currently work in Chrome, but won't work in the future, when things become partitioned based on the top-level frame's origin.

I think that, for the purposes of the Web Almanac we shouldn't consider this technique. At the end of the day, if sites register service workers for 3rd party origins we can assume that those origins might have a call to `navigator.serviceWorker.register` at some part of the site, so those would be eventually counted if we continue detecting registrations in that way.

Also good news, @jeffposnick will join this team as a reviewer as well 🎉

**Comment by bazzadp** [comment]
> Also good news, @jeffposnick will join this team as a reviewer as well 🎉

Can you add him to the meta data at the top of this issue then? It's used to keep our [chapter tracker](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2179) up to date so if you can remember to make any changes there that would be much appreciated!

**Comment by demianrenzulli** [comment]
I just added the new analysts there. Thank you!

**Comment by obto** [comment]
@demianrenzulli @webmaxru @Schweinepriester @thepassle @hemanth @tropicadri @andreban @jeffposnick @bazzadp

Hey everyone, wanted to check in and give you all a few reminder as we close in on our second deadline:

1. Looks like most of you already have, but for those who have not, please make sure you've requested edit access to the [Google Planning Doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit) and added your name to the first page as a reviewer, author or analyst.

2. Reminder that we need to have the chapter outline complete by **June 15** so we have enough time to update our crawler with any additional metrics you need this year. Seeing some good thoughts, ideas and suggestions so far. Looking forward to seeing the finished outline!

3. The team has a channel on slack (`#web-almanac-pwa`), so feel free to join everyone there as well: https://join.slack.com/t/httparchive/shared_invite/zt-45sgwmnb-eDEatOhqssqNAKxxOSLAaA

If you have any other questions don't hesitate to reach out :)

**Comment by demianrenzulli** [comment]
Thanks for the note @obto! 

I just wanted to share an update with this group from a conversation we had in the morning with @bazzadp, who kindly helped me get up-to-date with the challenges of this section for 2021:

- As Barry mentioned, we might not be able to reuse the `httparchive.almanac.service_workers` and `httparchive.almanac.manifests` tables that were used for the [2020 PWA queries](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2020/pwa).
- @rviscomi created a [new script](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/pwa.js) that might work well as a replacement, but we need to be sure that the regexes used there are good enough to detect service worker registrations and Manifests. That's why Barry asked the questions before, which I replied [here](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2153#issuecomment-849760433).

With that being said, our first immediate goal is to have a query migration plan that would allow us to give continuity to what was done in 2019 and 2020, and not depending in the old service worker and manfiest tables, if possible.

To start working on that, I just shared with this group [a sheet](https://docs.google.com/spreadsheets/d/1y-1f9Feq76fbBLBOON9yzK34f9M05mOfOLW4D6DlEE4/edit#gid=0) I put together with the summary of the data presented in 2020, and how (I think) we could migrate them to the new script.

I'm still ramping up in all the Almanac stuff, so I'm not 100% that my assumptions are correct, and if my understanding of Rick's script is accurate, so feel free to leave any comments there, or here, so we can iterate.

I hope this helps!

**Comment by demianrenzulli** [comment]
Hi again folks, one final update before the weekend:

We have been testing [Rick's script](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/pwa.js) to see if it works well.

So far, it seems like it does. Here is [an example test](https://webpagetest.org/custom_metrics.php?test=210528_BiDc2K_98c93c30b801e6710109c0269c5a788b&run=1&cached=0) for a PWA that uses Workbox if you want to take a look at how the **"PWA"** custom metric looks like for a site. 

One of our concerns was if the Workbox part of the script would detect all the Workbox modules written inside `workbox-sw.js` or only those that the library actually requests. In the example we shared, there are 8 Workbox modules listed and it seems that all of them are being used by the site's SW (which is fine). Comparing it with other sites, it seems like it's correct.

We'll be making 2 changes in the script to support the missing information:

- A regular expression matching all the possible events inside a service worker, to replace the [2020 SW Event Query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/pwa/sw_events.sql).
- Another regular expression to match the calls to `importScript()`, so we can replace the [2020 query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/pwa/popular_pwa_libraries.sql) used to create a ranking of popular libraries used by service workers.

Besides that, the analysis I shared [in the sheet](https://docs.google.com/spreadsheets/d/1y-1f9Feq76fbBLBOON9yzK34f9M05mOfOLW4D6DlEE4/edit#gid=0), seems to indicate that we should be ok.

If folks in this thread want to share any new metrics to capture this year, they are highly welcome. We might still have time to incorporate those!

**Comment by demianrenzulli** [comment]
Hi again folks, 

We have been working on this with @bazzadp, and our conclusion is that we'll be able to get the same data as 2020 from [Rick's new PWA script](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/pwa.js) 😀.

I submitted [a PR](https://github.com/HTTPArchive/legacy.httparchive.org/pull/204) to add some additional ones to detect calls to `importScripts()`, service worker events and service worker properties.

With that, we should be at least at parity with last year's metrics.

Our next goal is to define which new metrics we are going to use and, for that, we would need everyone's thoughts and ideas.

@obto: would it be possible to organize a follow-up call to discuss this?

If you already have things that would like to measure, please, feel free to share your thoughts.

Cheers.

**Comment by obto** [comment]
@demianrenzulli The chapter outline is due by June 15th, which is just over a week from now and it doesn't look like anything has been written as of yet. How can we help?

**Comment by demianrenzulli** [comment]
HI @obto,

I put together [this deck](https://docs.google.com/presentation/d/1_Eah-BWhUMpVUAs3ZiCElJrlFcVIzrhN12TWSNRbR8E/edit#slide=id.gdebe9d77dd_0_3), which contains the whole plan for the PWA chapter of Web Almanac in 2021.

The chapter outline starts in slide 17.

The reason that I did it in a deck format is because it will be easier to present and brainstorm with other contributors. I'll move that part of the deck to the doc now, so we can keep it there too.

**Rest of the team:** we'll be having a meeting tomorrow to discuss any new metrics you might have in mind, so, please bring your ideas. I marked in green in the deck the parts that will be different from 2020.

For those that can't make it to the call, feel free to share any thoughts you might have. I believe I gave permissions to all of you, but if anyone can't open it, please, request it, so I can give permission too.

**Comment by demianrenzulli** [comment]
I moved the outline to [the doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit#) as well. 

I'm working on a new set of metrics for [pwa.js](https://github.com/HTTPArchive/legacy.httparchive.org/blob/master/custom_metrics/pwa.js) and will update [the PR](https://github.com/HTTPArchive/legacy.httparchive.org/pull/204) soon.

@bazzadp: I'd like to finish the round of meetings with collaborators this week first, so we can discuss and finalize the list of metrics to include by end of week or the next.

**Comment by tunetheweb** [comment]
Hey @demianrenzulli , just an FYI I've changed my GitHub username to @tunetheweb to match my Twitter handle.

Let me know when your meetings are and can try and attend when I can. Otherwise you can represent the analysts and ping me separately.

**Comment by demianrenzulli** [comment]
HI @tunetheweb! We have a meeting tomorrow with the rest of the team. I might have forgotten to include you, since I just copied the kick-off invite and you were not yet part of the PWA team at that point.

Please, let me know if that works for you. If not, I would prefer to reschedule to a time that works. I think your presence would be super important at this stage.

It might be hard to follow the flow of the deck, but I think that format will definitely help us if I present it during a call.

**Comment by demianrenzulli** [comment]
For folks in this thread interested on following more technical discussions around the measurement techniques we'll be using and the latest  metrics, this is mostly taking place in [the PR](https://github.com/HTTPArchive/legacy.httparchive.org/pull/204) I shared before.

Feel free to share any ideas either here or as a comment in that PR directly.

**Comment by logicalphase** [comment]
I really like the presentation. It's polished and professionally formatted.
Thank you for sharing it with us. 💯

**Comment by christianliebel** [comment]
Hi @demianrenzulli! I'm the content lead for the Capabilities chapter (#2152). Since our two chapters are definitely related, I wanted to ask if we could find appropriate points to reference each other. Here are the chapters from last year: [Capabilities (2020)](https://almanac.httparchive.org/en/2020/capabilities) [PWA (2020)](https://almanac.httparchive.org/en/2020/pwa)

**Comment by demianrenzulli** [comment]
HI @christianliebel! Thank you for reaching out.

I think both chapters are definitely complementary, and we might have some overlap with some metrics, but I think that's not a problem. I'd be more than happy to link to the capabilities section as part of the introduction to PWA, when we'll mention the "App-like" features a PWA can have.

Regarding specific APIs, taking a look at last year's chapter, I see two features that we'll be covering in PWA as well:

- [Periodic Background Sync](https://almanac.httparchive.org/en/2020/capabilities#periodic-background-sync-api): We are detecting calls to that API. Since some of it's main use cases are related to caching.
- [Integration with Native App Stores](https://almanac.httparchive.org/en/2020/capabilities#integration-with-native-app-stores):  We are planning to cover this in depth in the PWA section, but it really depends on how much data app stores are willing to publish externally.

In both cases, I don't think it hurts to have both APIs covered in both chapters, as their intentions are different.

Let us know what you think. Happy to make a call with the capabilities team also, if you need to align more.

**Comment by webmaxru** [comment]
Hello folks! Status update: I'm in dialogue from MS Edge and PWA Builder teams. Hopefully, we'll get some interesting data about installability from their side.

**Comment by christianliebel** [comment]
> Regarding specific APIs, taking a look at last year's chapter, I see two features that we'll be covering in PWA as well:

@demianrenzulli Our focus this year is on productivity apps on the desktop, so we will discuss different APIs this year. If you want, you can check out our [2021 outline](https://docs.google.com/document/d/1pRk8FvCUWZbnaHc4vCQHbEHguNL5A0O34euJRIeUXM8/edit#heading=h.vqdmpb884vdu).

> In both cases, I don't think it hurts to have both APIs/features covered in both chapters, since each narrative will different and will have different objectives.

I had a look at your 2021 outline and I think the chapters don't overlap at all, so we shouldn't run into any problems here.

> Let us know what you think. Happy to make a call with the capabilities team also, if you need to align more.

Sure, let's schedule a quick call (15—30 min) to define where we would refer to each other's chapters (I've already added a potential position to our outline). I think a 1:1 conversation between us should be sufficient, we don't need to loop in the whole team. Would [Monday, Jun 28 2021, 09:00 (PDT) / 12:00 (EDT) / 17:00 (BST) / 18:00 (CEST)](https://www.timeanddate.de/zeitzonen/zeitzonenrechner?iso=20210628T160000&p1=964) work for you?

**Comment by demianrenzulli** [comment]
Thanks @christianliebel.

The time you proposed works for me. For simplicity, I don't think we need to include the rest of the team. I just scheduled a call with them the next day, where I can also share the outcomes of this call.

Please, feel free to send the invite to the email that appears [in the doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit). I can schedule one otherwise.

Cheers.

**Comment by demianrenzulli** [comment]
Hi folks,

I just pushed [my first PR](https://github.com/HTTPArchive/almanac.httparchive.org/pull/2252) for a small content change to the 2020 Almanac. It helped me [setting up the environment](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/src), which I highly recommend to anyone that wants to contribute at some point.

While doing the change, I've realized that some chapters contain translations, which is something that developers have asked for. We recently did a project called [International Mother Language Day](https://developer.chrome.com/blog/mother-language-day-2021/), which was well received by the community and were thinking on different ways of delivering the content in different languages.

I'm wondering how important do you think having the content translated could be for the developer communities of your countries / regions? If you think it's valuable, maybe we can work on the translations and share in non-English speaking countries:

- @tropicadri and me in Spanish
- @andreban in Portuguese
- @Schweinepriester in German
- @webmaxru: From your profile I assume it's Norwegian (please, correct me if I'm wrong).

Cheers.

**Comment by tunetheweb** [comment]
Thanks for raising awareness of the translations @demianrenzulli ! We think they’re important to make the content as inclusive as possible.

Spanish and Portuguese are all set up so any chapter can be translated to those. German and Norwegian will be a bit more effort as the base templates are not translated yet but will we’re grateful for any translations if people want to contribute and happy to talk through the process. These can be set up in advance if you do plan on supporting those. And similarly if you wanna translate the 2020 PWA chapter (or other chapters) to get you used to it then definitely will be happy to accept that!

Each language has its own issue [linked from here](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/2020-Translation) and we’ll add the 2021 chapters to those issues when they are written. We ask people to put their name next to a chapter in that languages’s issue to let people know you’re working on that so others don’t duplicate work.

**Comment by demianrenzulli** [comment]
Thanks @tunetheweb!

Regarding this:

> Spanish and Portuguese are all set up so any chapter can be translated to those. German and Norwegian will be a bit more effort as the base templates are not translated yet but will we’re grateful for any translations if people want to contribute and happy to talk through the process.

I thought that contributors would just take a chapter and translate them "manually". From your previous comment, it seems like there's an automatic process for that instead? Please, correct me if I'm wrong, or If you have a link to share about the process, that might be of help.

**Comment by tunetheweb** [comment]
No you are correct. However, as well as the chapter's markdown text there's a bit of scaffolding to create the the headers and footer, the home page, the contributors page...etc. to allow us to run the whole site in that language. And these have to be done before we can accept a chapter translation.

As I say they have been done for Spanish and Portuguese (have a look at [src/temples/es](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/src/templates/es) for example), but not for German or Norwegian. And if you look at #498 you can see these templates have all been "ticked" for Spanish to say they are done, so just chapters left to do really, whereas all unticked for German in #1300.

Get me?

**Comment by demianrenzulli** [comment]
Got it! It seems like we can, at the very least, ship the chapter in English and Spanish. Portuguese could be also an option, as long as @andreban has the time to help translating the content to Portuguese and thinks it might be valuable.

Regarding the rest of the languages, I don't want to assume that folks here will have the bandwidth to work on that, since those seem to involve some additional work, but I think it's useful to know that there are pending tasks that anyone can take if they wish to contribute to those languages at some point.

As mentioned before, some communities might feel more comfortable with English content than others. From recent interactions in the Spanish/Portuguese speaking dev communities, it seems like there's appetite for content in their own languages, but I can't speak for the rest.

This is an interesting aspect to address, that I believe most folks here would be interested in.

**Comment by tunetheweb** [comment]
Good stuff. More info here btw on the process: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Translators'-Guide

And similar wiki pages in there on Authoring and Analysing and the other roles if you've not seen them yet.

**Comment by tunetheweb** [comment]
@rviscomi / @demianrenzulli I've been working on converting all the 2020 queries to custom metric versions and making good progress on that. However I'm a little concerned on the number of service workers we're missing and can't explain why, as some of them should have be caught by our custom metric.

For example running the below custom metric on https://www.mediamarkt.at/ returns no service workers as measured by our custom metric (the top bit), but does return two registration scripts from a simpler search (`swRegistrationCount` lines near the bottom), and I can clearly see a service worker being registered in the _main.js file.

Any ideas what's going on here, as not having much luck figuring it out, despite the fact this easily repeatable on WPT. Then again the mere sight of `reduce` makes me run away screaming which isn't helping me to figure out the issue...

```js
[sw]
let response_bodies = $WPT_BODIES;
const serviceWorkerRegistrationPattern = /navigator\.serviceWorker\.register\(['"]([^"']+)/m;

const serviceWorkerURLs = response_bodies.filter(har => {
  return serviceWorkerRegistrationPattern.test(har.response_body);
}).map(har => {
  const base = new URL(har.url).origin;
  const serviceWorkerPath = har.response_body.match(serviceWorkerRegistrationPattern)[1];
  return new URL(serviceWorkerPath, base).href;
}).reduce((set, url) => {
  set.add(url);
  return set;
}, new Set());

function getEntriesForURLs(urlSet) {
  return response_bodies.filter(har => {
    return urlSet.has(har.url);
  }).map(har => {
    return [har.url, har.response_body];
  });
}

const serviceWorkers = getEntriesForURLs(serviceWorkerURLs);

let swRegistrationCount = 0;
response_bodies.forEach((file) => {
        swRegistrationCount += ((file.response_body || '').match(serviceWorkerRegistrationPattern) || []).length;
    }
)
return {
  serviceWorkers: Object.fromEntries(serviceWorkers),
  swRegistrationCount: swRegistrationCount
}
```

**Comment by tunetheweb** [comment]
OK I've figured it out. The custom metric has the following:

```js
  const base = new URL(har.url).origin;
  const serviceWorkerPath = har.response_body.match(serviceWorkerRegistrationPattern)[1];
  return new URL(serviceWorkerPath, base).href;
```

That is it takes the service worker URL, and prepends the origin of the registering script if it's a relative path.

The problem is for scripts served from asset domains.

So https://www.mediamarkt.at/
calls https://asset.mediamarkt.at/webmobile-pwa-public-assets-dev/b4d5e5d/6479d5f494219c2cef13_main.js
which then calls:
```js
 navigator.serviceWorker.register("/sw.js").then((function() {
                        console.log(e, "Service Worker registered successfully")
                    }
```
which the current custom metric converts into the URL https://asset.mediamarkt.at/sw.js
instead of https://www.mediamarkt.at/sw.js
which then doesn't make any URL requested, so the serviceWorker object is empty :-(

I guess we could convert relative URLs to be relative to a base of `window.origin` instead `new URL(har.url).origin`, though that wouldn't work for iframes with relative path URLs. Probably more important to get main window service workers though, than iframes.

But what to do for the 2021 Web Almanac? Use this data or revert to `response_bodies` like we did last year (assuming we can get that)

Looking at the June data, we found 70,260 service workers, while Lighthouse reckons there were 107,754. We always knew there was going to be a discrepancy as we're doing fair basic checks, but being down by about a third is more than I thought we'd be to be honest.

However most of the other metrics should still work - it's mainly the `serviceWorker` metric itself that's a bit flawed. However it does make it more difficult to filter some queries to service worker pages only (unless we do an expensive join to Lighthouse or create a temp table of that to make it cheaper).

But we may have a workaround for that too as @demianrenzulli also added a `serviceWorkerHeuristic` query, which uses a few other fields, on top of `serviceWorkers` field to "guess" if it's a ServiceWorker:

```js
//Experimental field: Heuristic to detect if a site has a service worker even if the 'serviceWorkers' field is empty (false positives).
  serviceWorkerHeuristic: !isObjectKeyEmpty(serviceWorkers) || !isObjectKeyEmpty(workboxInfo) || !isObjectKeyEmpty(swEventListenersInfo) || !isObjectKeyEmpty(swMethodsInfo)
```

This looks to work for this example, and I'm guessing it should for most of the others, but as it wasn't on for the whole of the June crawl it's difficult to be sure.

So I'd suggest we do the following:
1. Update the Custom Metric to look at `window.origin` for next year.
2. Use the `serviceWorkerHeuristic` to define if it's a service worker page and see how far that is out compared to Lighthouse after July finishes.
3. Revert back to `response_bodies` if we really have to (or maybe run it one last time anyway as a double check).

What do you think?

**Comment by tunetheweb** [comment]
> Update the Custom Metric to look at `location.origin` for next year.

Or alternatively we change the custom metric to log the `serviceWorker` usage always - even if it doesn't match an exact URL in the har, but just don't get the service worker response body in the custom metric in that case?

**Comment by tunetheweb** [comment]
I did some more analysis on this as shown in the table below:

serviceworkers | 70,260
-- | --
manifests | 534,332
serviceWorkerHeuristic | 62,344
serviceWorkerHeuristicFalse | 2,023,514
serviceWorkerHeuristicNull | 5,347,730
workboxInfo | 56,605
swEventListenersInfo | 59,214
swMethodsInfo | 38,417
lh_uses_sw | 107,754
lh_true_heuristic_false | 429
lh_false_heuristic_true | 32,822
lh_false_wb_true | 20,736
lh_false_sweventListeners_true | 30,467
lh_false_swmethods_true | 17,666
total | 7,447,627

What this is saying is that:
- We identified 70,260 sites as using service workers using our custom metric, which under counts the above scenario, which should be for the full run.
- Lighthouse identified 107,754 sites (again for full run)
- Our heuristic identified 62,344 sites as service workers for a partial run, and most of the run (5,347,730 sites - or 71% of the run) had no heuristic as it was only added late in the run.
- There were 429 sites where Lighthouse said there was a service worker, but we said there was not. That's reassuringly low, even if it is only based on partial run data.
- There are however 32,822 sites where we think they have a service worker (according to the heuristics) and Lighthouse failed the service-worker audit. Now that's a good bit higher than we'd like but the (Lighthouse Service Worker Audit](https://web.dev/service-worker/) does more than just check for the presence of a Service-Worker, but also insists other criteria are met (a manifest, with a `start-url` property) is present, and the service-worker controls the top level page - so not unexpected that we're higher. Though still a bit higher than I'd like!
- Looking at the breakdown of the heuristics it seems across the board with workbox usage, serviceworker events, and service worker methods all registering significant numbers even when lighthouse fails the audit. As these should have relatively low false positives, this to me suggests it is perhaps iframes or the `start_url` requirement failing, so (arguably) they are service-worker controlled pages as far as our analysis is concerned, our analysis so maybe we are comfortable with this gap?
 
And maybe we can cover in writing when defining what pages we looked at (those with any service-worker usage and a manifest), and also by presenting the Lighthouse data and noting it's stricter definition of a PWA?

<details>
<summary>SQL Query to get above results (warning uses 3TB)</summary
<code>
<pre>
#standardSQL

CREATE TEMPORARY FUNCTION getAudits(report STRING, category STRING)
RETURNS ARRAY<STRUCT<id STRING, weight INT64, audit_group STRING, title STRING, description STRING, score INT64>> LANGUAGE js AS '''
var $ = JSON.parse(report);
var auditrefs = $.categories[category].auditRefs;
var audits = $.audits;
$ = null;
var results = [];
for (auditref of auditrefs) {
  results.push({
    id: auditref.id,
    weight: auditref.weight,
    audit_group: auditref.group,
    description: audits[auditref.id].description,
    score: audits[auditref.id].score
  });
}
return results;
''';

SELECT
  SUM(ServiceWorker) AS serviceworkers,
  SUM(manifests) AS manifests,
  SUM(serviceWorkerHeuristic) AS serviceWorkerHeuristic,
  SUM(serviceWorkerHeuristicFalse) AS serviceWorkerHeuristicFalse,
  SUM(serviceWorkerHeuristicNull) AS serviceWorkerHeuristicNull,
  SUM(workboxInfo) AS workboxInfo,
  SUM(swEventListenersInfo) AS swEventListenersInfo,
  SUM(swMethodsInfo) AS swMethodsInfo,
  COUNTIF(score > 0) AS lh_uses_sw,
  COUNTIF(score > 0 AND serviceWorkerHeuristicFalse > 0) AS lh_true_heuristic_false,
  COUNTIF(score = 0 AND serviceWorkerHeuristic > 0) AS lh_false_heuristic_true,
  COUNTIF(score = 0 AND workboxInfo > 0) AS lh_false_wb_true,
  COUNTIF(score = 0 AND swEventListenersInfo > 0) AS lh_false_sweventListeners_true,
  COUNTIF(score = 0 AND swMethodsInfo > 0) AS lh_false_swmethods_true,
  COUNT(0) AS total
  /*
  lh.url,
  pwa_payload
  */
FROM (
  SELECT
    url,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkers') != "[]",1,0) AS ServiceWorker,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = 'true',1,0) AS serviceWorkerHeuristic,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = 'false',1,0) AS serviceWorkerHeuristicFalse,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') IS NULL,1,0) AS serviceWorkerHeuristicNull,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.workboxInfo') != "[]",1,0) AS workboxInfo,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.swEventListenersInfo') != "[]",1,0) AS swEventListenersInfo,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.swMethodsInfo') != "[]",1,0) AS swMethodsInfo,
    IF(JSON_EXTRACT(payload, '$._pwa') != "[]" AND JSON_EXTRACT(payload, '$._pwa.manifests') != "[]",1,0) AS manifests,
    JSON_EXTRACT(payload, '$._pwa') as pwa_payload
  FROM
    --`httparchive.sample_data.pages_mobile_10k`
    `httparchive.pages.2021_06_01_mobile`
  ) p
LEFT OUTER JOIN (
  SELECT
    url,
    audits.score as score
  FROM
    --`httparchive.sample_data.lighthouse_mobile_10k`,
    `httparchive.lighthouse.2021_06_01_mobile`,
    UNNEST(getAudits(report, "pwa")) AS audits
  WHERE
    audits.id = 'service-worker' AND
    LENGTH(report) < 20000000  # necessary to avoid out of memory issues. Excludes a handful of very large results
) lh
ON p.url = lh.url
</pre>
</code>
</details>

**Comment by demianrenzulli** [comment]
Hi folks,

Thanks again for all the great work @tunetheweb! We are starting to understand the limitations of our new approach better since you shared all analysis.

We have been talking offline about this, and here is our decision so far:

- The fact that the `serviceWorkers` has so many false positive is not an issue for the **Service Worker** section itself: we'll use the `blink_usage` table to determine the % of URLs that use service workers and [Chrome Platform Status](https://www.chromestatus.com/metrics/feature/timeline/popularity/990) for the % of page loads.
- Most of the scripts to detect usage of service worker features (e.g. SW methods) don't use the `serviceWorkers`  field, so they shouldn't be impacted by this.
- The one that might be impacted is the analysis of how many sites have service workers AND manifests. For that we use either:

1. the `serviceWorkers` stat
2. The `serviceWorkerHeuristic` field (if we determine it's better).
3. Lighthouse (expensive and slow)
4. Response_bodies (expensive and slow)

With that said, it seems like even when the `serviceWorkers` stat doesn't work as expected, we should be fine.

@rviscomi: regardless of this analysis, it seems like [this comment](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2153#issuecomment-874570418) from @tunetheweb points to a bug. Do you agree with that? In case you do, you think you could apply a quick fix before the next crawl?

That seems to be the only AI.

**Comment by tunetheweb** [comment]
> @rviscomi: regardless of this analysis, it seems like [this comment](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2153#issuecomment-874570418) from @tunetheweb points to a bug. Do you agree with that? In case you do, you think you could apply a quick fix before the next crawl?

Looking at the previous [pwa_candidates table](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/util/pwa_candidates.sql) it uses the `page` URL as the base, and not the service worker js origin, so they wouldn't have covered relative iframes either. So think we are good to change this for future years.

Raised https://github.com/HTTPArchive/legacy.httparchive.org/pull/223 and can merge and release after this crawl, and before next month.

**Comment by webmaxru** [comment]
Hello! This is the data for our report I received from the PWA Builder team:

On the 200K last URLs reviewed (3 months’ timeslot) by PWA Builder:

- 75% had a manifest detected
- 11.5% had a service worker detected
- 9.6% are installable PWAs from the browser (manifest+SW+https)

@demianrenzulli could we add this section to the overall report?

**Comment by demianrenzulli** [comment]
Thanks for sharing these stats @webmaxru! 

**Everyone:** We have working with @tunetheweb on the queries and analyzing the data and I have started to write the chapter in the [shared doc](https://docs.google.com/document/d/1grNpHfQgBZow7fTw9UNZk8E4eWdQV97ZyenPmaye1JU/edit#). I was thinking asking reviewers to help by the end of week, when it will be almost finished.

@webmaxru: I have added a placeholder for a "Distribution" section at the end of the chapter, where we can talk about different ways to distribute PWA, including tools like PWA builder, bubblewrap, etc.

I have just added an entry for "PWA Builder" and included your stats. there I might ask you to write a summary of the tool and what it is for, so readers that are not familiar know more a about it. 

Let me know what you think.

**Comment by tunetheweb** [comment]
Do we have a source for these stats @webmaxru ? Or at least a link? We try to be as transparent as possible where our data comes from (e.g. we publish SQL queries for the HTTP Archive data) so should try our best to be similarly transparent (where possible) for external stats too.

**Comment by rviscomi** [comment]
I was literally typing out a message right now to say the same thing 😄 

+1 to having a public source for the stats, like an official blog post with a methodology if possible.

**Comment by webmaxru** [comment]
Will reach out to David Rousset from PWA Builder team (who shared the data with me) with your requests, folks.

**Comment by davrous** [comment]
Hi! [PWA Builder](https://www.pwabuilder.com) is an open-source project that can help web developers to build Progressive Web Apps (PWAs) and package them for app stores (Microsoft Store & Google PlayStore). It starts by reviewing a provided URL to check for an available manifest, service worker and SSL. 

We have a private reporting API simply collecting the URLs submitted on our home page and attached to it the score as well as the PWA features detected: manifest available or not, service worker registered or not, valid SSL available or not. 

Based on that, I've done an internal query to get the last 200K URLs reviewed to compute the results shared above. Our Reporting API is not publicly available.

**Comment by rviscomi** [comment]
Thanks @davrous and @webmaxru! Would it be possible for you to publish a lightweight blog post at blog.pwabuilder.com with this info/data? That would give us a nice authoritative/linkable source to reference in the Web Almanac chapter.

**Comment by webmaxru** [comment]
Hello! Just a day ago I found another source of PWA stats:
https://mobsted.com/world_state_of_pwa_2021
"It took us about 9 months to deep parse about 300 million top-level domains and billions of connected web pages to learn the current state of Progressive Apps in the World Wide Web."

I know Alex Burlitsky from Russia who is behind this research. Do we want to also include this data in the report?

**Comment by alexburleh** [comment]
Hi all. It would be great if our research helps more people to move to our favourite tech. My team invested lots of resources to make it happen. We started with a db of 300 million websites with a self made bots network. First looked for signs of PWA going 5 links down from every website, then returned for deeper look to study manifests or workers are valid. We set values depending on the state of each parameter. All those where PWAs were validated went through addition checks. We recorded languages and keywords, then used external APIs to determine industry of the website, traffic and search positions. We also followed the pwa making funnel. The report which @webmaxru mentioned here holds the high level data from this study.

**Comment by demianrenzulli** [comment]
Thanks for sharing this analysis @webmaxru and congratulations to @alexburleh and his team. The research looks really interesting.

Taking a look at it, I think there's an opportunity to reference the **"Use cases by industry. Where PWAs are applied best?"** part of the analysis in one section of our chapter. 
Last year's chapter included  ["Top Manifest Categories"](https://almanac.httparchive.org/en/2020/pwa#top-manifest-categories), but we have found that the property is being used in very few sites (less than 0.8%), so we won't be including it this year. We could then reference this analysis and one of the sources that contain usage data by industry.

@alexburleh: do you want to provide more details on how your team came up with those particular numbers?

Thanks.

**Comment by webmaxru** [comment]
@demianrenzulli, I'm also in dialogue with Microsoft Edge team about PWA installation/usage data they can share. For now, public-ready statements are:
- Millions of monthly installs
- Billions of monthly usage minutes

I'll check if we can concretize/extend these statements.

**Comment by alexburleh** [comment]
Hi @demianrenzulli , sure. Do you want it here for understanding, or do you need like a "ready made" text explaining how we got there? 

If for understanding  here it is. For those 570+ thousand confirmed progressives, we took texts at random from different pages of each and ran it through Google's cloud recognition API to determine the industry. Then we counted only ones with over 75% confidence rate, it was over 90% of the initial batch. Then counted total industries where at least one pwa is found (574). You can see two charts up there, one is bar chart with the high level groupings, like Business and Industries, Hobbies, etc etc. To show were are most apps found. The other is pie chart which has every one of the higher level ones broken down in separates, so hobbies are split into hikeing, camping, etc etc. This way we can see that usage is really diverse.

**Comment by demianrenzulli** [comment]
Thanks @webmaxru! It would be great to count with those numbers. I'm also trying to pull some stats from Chrome regarding .apk installations.

@alexburleh: This is great! I'll use your explanation in that section (categories), and link to your analysis so readers can continue reading the rest of the research there 👍

**Comment by alexburleh** [comment]
Sure, @demianrenzulli and thanks! Don't hesitate to shoot any other questions.

**Comment by alexburleh** [comment]
@demianrenzulli 

0) very interesting data :))) 

1) As an idea - I think it would be easier for users to see the chart on Industries. Mine are of different designs and colors. So if you agree I can either send you aggregate raw data to add the chart in your designs, or I can pick your colors from the page and send you a ready-made.

2) on page 17 there is a Section - called "popular import scripts". Is this something that relates to our product called "PWA JS Capsule" - https://docs.mobsted.com/display/KB/3+Ways+to+Convert+-+how+to+choose - its the right most option. Projects only add a JS and 2 files of workers to their site and it gets installable, with push, custom prompts and all that. Last time I checked there are about 500 projects who've chosen this way, and we released it only this year.

3) The other type of PWA-less PWA is our Wrapper (same link above), you mention tools like Builder and Bubblewrap, but skip this method of creating apps, which also is used by many websites, who have no access to the CMS or code, say when built on Wix :)

4) At first I was wondering why there is a difference between your data and ours. As you point to 8 times more active apps (yours 1.71 and ours is 0.2). Then I conclused that it may have something to do with how we check installability. Then you show that 61.73% of PWA sites have an Installable Manifest, which cuts our discrepancy to like 4X. To check installability we not only see the lighthouse but actually raise the browser session and view results. Maybe that's the explanation.

**Comment by demianrenzulli** [comment]
Thanks so much to all the reviewers for sharing your thoughts in the doc so promptly! I'm giving some additional time, so all the participants that are interested can share theirs, and then will start working on the comments.

Also, @alexburleh: thanks for sharing your suggestions!

Regarding data/measurement: it might differ with yours  in the way we are measuring things and also in the number of URLs we are analyzing (I believe you took a much bigger sample). In any case, your hypothesis around installability and manifests seem to make sense to me. Let's keep in mind that this is mostly an approximation, and that the goal is that, if readers disagree with our assessment, they can reproduce the same results by running [the queries](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2021/pwa) themselves, and maybe modifying some of the conditions we used, to get results that are closer to what they think it's correct.

Regarding the industry chart: do you want to provide it by pasting it in the doc, or maybe you can share it here and I can dd it to that section? Just let us know. I'm not very sure about the libraries part, since that data comes from a query, and we still need to discuss with @tunetheweb if it makes sense to continue using it this year.

**Comment by alexburleh** [comment]
@demianrenzulli attached, only the chart, I could not get the header font right 🤷‍♂️

Suggest a header "Share of PWAs by Industry" 
3 different colors are just leaders for design purpose, no more.
<img width="657" alt="chart_pwa_by_industry_2021" src="https://user-images.githubusercontent.com/91685593/136053828-a2f5019d-9af1-4075-8618-23e8d975b125.png">

**Comment by demianrenzulli** [comment]
That's awesome! Thanks so much @alexburleh. I'll add that chart to the article doc, along with the reference to your analysis👍

**Comment by rviscomi** [comment]
@alexburleh any chance you could share the chart data in CSV format so we could run it through our data viz process for visual consistency with other charts?

@demianrenzulli @tunetheweb we should be sure to annotate the chart with `Source: Mobsted PWA 2021 report` in the subtitle.

**Comment by tunetheweb** [comment]
Any chance we could get the raw data instead of the chart @alexburleh ? That way we can use our own chart style (appreciate you got it very close, but still slightly different and those stylist highlights might look odd).

And also means we can make it interactive on desktop like the other charts (not that interactive to be fair - just hover effects, but still!)

Also FYI, in your report you have a typo in the header:

![image](https://user-images.githubusercontent.com/10931297/136058309-d32e8ea9-3ecf-477d-8307-f19b8d8ac6a5.png)

**Comment by alexburleh** [comment]
@rviscomi  @tunetheweb  Gentlemen, 1) attached is the CSV for the above diagram 2) Barry, thanks for the typo, will fix that :)
[PWA agregbyindustry.csv](https://github.com/HTTPArchive/almanac.httparchive.org/files/7287629/PWA.agregbyindustry.csv)

**Comment by tunetheweb** [comment]
Thanks @alexburleh - what's the total number to convert to %?

**Comment by alexburleh** [comment]
Those is csv are the % already, which all add up to 100. The total
researched number was ~570K if I remember correctly.

**Comment by tunetheweb** [comment]
Ah comma delimited - you crazy Continental Europeans 😁

**Comment by alexburleh** [comment]
Don't blame me :) I just know how to download the csv out of Google sheets
file :)

**Comment by tunetheweb** [comment]
Added to sheet: https://docs.google.com/spreadsheets/d/16AkIdDBBkCR5Kgb7kyfYvnNLQBu23Vsh7MUSFHW9RtA/edit#gid=1817367668

What do you all think of this as a visualisation?:

![Bar chart of PWA Categories](https://user-images.githubusercontent.com/10931297/136072717-67937490-a86c-46e7-8dbf-3cb9512b1b4e.png)

**Comment by demianrenzulli** [comment]
This looks very aligned to the way the rest of the charts look like. Thanks a lot for working on this @tunetheweb!

**Comment by alexburleh** [comment]
For some reason the chart deviates from raw data. For example Biz and
Industrial should be 14%, not 1.4 and Computers and Electronics should be
2.1, not 0. Sensitive subjects should be 0.9, not 9.3. Online communities
should be less then 1%, not 7.6

I may have missed something too.

**Comment by tunetheweb** [comment]
Oops, conversion error. Take 2:

![image](https://user-images.githubusercontent.com/10931297/136074790-675fe84b-01b5-4fb3-9963-a7ed2e6de1ab.png)

**Comment by alexburleh** [comment]
> Oops, conversion error. Take 2:
> 
> ![image](https://user-images.githubusercontent.com/10931297/136074790-675fe84b-01b5-4fb3-9963-a7ed2e6de1ab.png)

Good. My only suggestion would be to color top 3 or 5 categories with one color and the rest with other. Could be easier to view.

**Comment by tunetheweb** [comment]
> Good. My only suggestion would be to color top 3 or 5 categories with one color and the rest with other. Could be easier to view.

Sorry that's not really the style we use so would prefer to be consistent here rather than introduce new themes. @demianrenzulli can could out particular stats in the writing if he wants.

**Comment by alexburleh** [comment]
Sure:) Your decision. It was just the suggestion. BTW just to make sure - I
meant not top 3 or 5 from top to bottom, but top 3 or 5 percentage wise :),
so Biz and Industries, Arts and Ent, Home & Garden etc.

Anyways guys, do as you need to :) I have no further Qs :)

**Comment by alexburleh** [comment]
Hi :)

It is probably a misprint in my little block -

Mobsted recently published THEY World State of PWA....  may be "their"?

**Comment by webmaxru** [comment]
@rviscomi @demianrenzulli @davrous : here is a tweet from PWA Builder to confirm stats from their side: https://twitter.com/pwabuilder/status/1454250060326318082?s=21

**Comment by tunetheweb** [comment]
The PWA 2021 chapter is completed and has been merged!

Well done all involved - especially @demianrenzulli for holding the ship together.

The first 2021 chapter merged! 🎉 Quite an achievement. Can't wait until it's launched.

@demianrenzulli @webmaxru @Schweinepriester @thepassle @hemanth @tropicadri @andreban @jeffposnick  @tunetheweb @rviscomi  @obto

**Comment by demianrenzulli** [comment]
Hi folks! 

Thanks again for collaborating in this year's PWA chapter!!

The chapter has already been released: https://almanac.httparchive.org/en/2021/pwa

Please, feel free to share or retweet the HTTP Archive announcement: https://twitter.com/HTTPArchive/status/1461029997804720131

cc // @webmaxru @Schweinepriester @thepassle @hemanth @tropicadri @andreban @jeffposnick

**Comment by webmaxru** [comment]
Thanks for leading the project, @demianrenzulli!
Please, keep me in mind for the next year's PWA chapter!

**Comment by demianrenzulli** [comment]
Sure thing! Thank you @webmaxru  and thanks for the great insights everyone, it was great working with you, and once again many thanks to @tunetheweb for his contributions, support (and patience)!!

@rviscomi has shared [this interest form](https://docs.google.com/forms/d/e/1FAIpQLSezsQmHufYjJqN1H18bUN8G8Qe1B-h02TMAlWVx6H7vWH0RtQ/viewform) so you can sign up and be contacted when the 2022 chapter starts.

**Comment by rviscomi** [comment]
@demianrenzulli @webmaxru @Schweinepriester @thepassle @hemanth @tropicadri @andreban @jeffposnick @tunetheweb

**Thank you** all for your hard work getting this chapter over the finish line in time for the pre-release—the PWA chapter has been the most-read chapter in the past few weeks, English and Russian translation combined! Congratulations on finishing the chapter, and I'm excited to see us launch the rest of the chapters along side it on Wednesday 🎉

When you get 5 minutes, I'd really appreciate if you could fill out our [contributor survey](https://forms.gle/VRzHPTyWxuwa67vF7) to tell us (the project leads) about your experience. It's super helpful to hear what went well or what could be improved for next time. 🙏
